### PR TITLE
Enable Optional Scope Validations feature for applications

### PR DESCRIPTION
**Pull request by HiranyaKavishani**
## Purpose
This feature is regarding the select optional scope validations for applications in APIM side.

## Approach
1. Enable optional scope validations in DefaultKeyValidationHandler.
2. Avoid the duplicate scope validations by disabling scope validation at IS side

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java [hunk 1]
```diff
@@ -170,32 +179,70 @@ public boolean validateScopes(TokenValidationContext validationContext) throws A
...
-                        if (log.isDebugEnabled()) {
-                            log.debug("Scope validation failed from "+ validator);
+                try {
+                    if (validator != null && appScopeValidators.contains(validator.getValidatorName())) {       //take the intersection of defined scope validators and scope
+                        if (log.isDebugEnabled()) {                                                             //validators registered for the application
+                            log.debug(String.format("Validating scope of token %s using %s", accessTokenDO.getTokenId(),
+                                    validator.getValidatorName()));
+                        }
+                        boolean isValid = validator.validateScope(accessTokenDO, resource);
+                        appScopeValidators.remove(validator.getValidatorName());
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java]: we can move this after the below if check I assume

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java [hunk 2]
```diff
@@ -170,32 +179,70 @@ public boolean validateScopes(TokenValidationContext validationContext) throws A
...
+                } catch (IdentityOAuth2Exception e) {
+                    log.error("ERROR while validating token scope " + e.getMessage(), e);
+                    apiKeyValidationInfoDTO.setAuthorized(false);
+                    apiKeyValidationInfoDTO.setValidationStatus(APIConstants.KeyValidationStatus.INVALID_SCOPE);
+                    return false;
                 }
             }
+
+            if (!appScopeValidators.isEmpty()) {   //if scope validators are not defined in identity.xml but there are scope validators assigned to an application, throws exception.
+                throw new IdentityOAuth2Exception(String.format("The scope validators %s registered for application %s@%s" +
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java]:
```suggestion
                throw new IdentityOAuth2Exception(String.format("The scope validator(s) %s registered for application %s@%s" +
```

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java [hunk 3]
```diff
@@ -170,32 +179,70 @@ public boolean validateScopes(TokenValidationContext validationContext) throws A
...
+                                " are not found in the server configuration ", StringUtils.join(appScopeValidators, ", "),
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java]:
```suggestion
                                " is/are not found in the server configuration ", StringUtils.join(appScopeValidators, ", "),
```

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/ExtendedDefaultTokenValidator.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    /**
+     * Validate scope of the access token using scope validators registered for that specific app.
+     *
+     * @param messageContext Message context of the token validation request
+     * @return Whether validation success or not
+     * @throws IdentityOAuth2Exception Exception during while validation
+     */
+    @Override
+    public boolean validateScope(OAuth2TokenValidationMessageContext messageContext) throws IdentityOAuth2Exception {
+        return true;    //scope validation happens in IS side always return true.
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/ExtendedDefaultTokenValidator.java]:
```suggestion
        return true;    //scope validation is handled in DefaultKeyValidationHandler
```

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/ExtendedDefaultTokenValidator.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
+package org.wso2.carbon.apimgt.keymgt.handlers;
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/ExtendedDefaultTokenValidator.java]: license header is misisng

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java [hunk 6]
```diff
@@ -170,32 +179,69 @@ public boolean validateScopes(TokenValidationContext validationContext) throws A
...
+        OAuthAppDO appInfo;
+        OAuthAppDAO oAuthAppDAO = new OAuthAppDAO();
+
         try {
-            Set<OAuth2ScopeValidator> oAuth2ScopeValidators = OAuthServerConfiguration.getInstance()
-                    .getOAuth2ScopeValidators();
-            //validate scope from all the validators
+            appInfo = oAuthAppDAO.getAppInformation(clientId);
+            scopeValidators = appInfo.getScopeValidators();     //get scope validators from the DB
+            ArrayList<String> appScopeValidators = new ArrayList<>(Arrays.asList(scopeValidators));
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java]: Use genetrice List<String> abc = new ArrayList

File: components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java [hunk 7]
```diff
@@ -170,32 +179,69 @@ public boolean validateScopes(TokenValidationContext validationContext) throws A
...
+
+            if (ArrayUtils.isEmpty(scopeValidators)) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java]: What happen to existing application which doens't have scope validators and we should support the default behavior.

Comment (HiranyaKavishani) [reply components/apimgt/org.wso2.carbon.apimgt.keymgt/src/main/java/org/wso2/carbon/apimgt/keymgt/handlers/DefaultKeyValidationHandler.java]: When the applications haven't been configured with any scope validators, it should return true as the default behavior.

**Comment by HiranyaKavishani** [comment]
moved to 6.x

**Pull request closed without merging**